src/test/fio/fio_ceph_messenger.cc

```diff
@@ -0,0 +1,702 @@
+ */
+
+#include "global/global_init.h"
+#include "msg/Messenger.h"
+#include "messages/MOSDOp.h"
+#include "messages/MOSDOpReply.h"
+#include "common/perf_counters.h"
+#include "ring_buffer.h"
+
+#include <fio.h>
+#include <flist.h>
+#include <optgroup.h>
+
+enum ceph_msgr_type {
+  CEPH_MSGR_TYPE_UNDEF,
+  CEPH_MSGR_TYPE_POSIX,
+  CEPH_MSGR_TYPE_DPDK,
+  CEPH_MSGR_TYPE_RDMA,
+};
+
+const char *ceph_msgr_types[] = { "undef", "async+posix",
+				  "async+dpdk", "async+rdma" };
+
+struct ceph_msgr_options {
+  struct thread_data *td__;
+  unsigned int is_receiver;
+  unsigned int is_single;
+  unsigned int port;
+  const char *hostname;
+  const char *conffile;
+  enum ceph_msgr_type ms_type;
+};
+
+class FioDispatcher;
+
+struct ceph_msgr_data {
+  ceph_msgr_data(struct ceph_msgr_options *o_, unsigned iodepth) :
+    o(o_), msgr(NULL), disp(NULL),
+    io_inflight_nr(0),
+    io_pending_nr(0) {
+    INIT_FLIST_HEAD(&io_inflight_list);
+    INIT_FLIST_HEAD(&io_pending_list);
+    ring_buffer_init(&io_completed_q, iodepth);
+    pthread_spin_init(&spin, PTHREAD_PROCESS_PRIVATE);
+  }
+
+  struct ceph_msgr_options *o;
+  Messenger *msgr;
+  FioDispatcher *disp;
+  pthread_spinlock_t spin;
+  struct ring_buffer io_completed_q;
+  struct flist_head io_inflight_list;
+  struct flist_head io_pending_list;
+  unsigned int io_inflight_nr;
+  unsigned int io_pending_nr;
+};
+
+struct ceph_msgr_io {
+  struct flist_head list;
+  struct ceph_msgr_data *data;
+  struct io_u *io_u;
+  MOSDOp *req_msg; /** Cached request, valid only for sender */
+};
+
+struct ceph_msgr_reply_io {
+  struct flist_head list;
+  MOSDOpReply *rep;
+};
+
+typedef union {
+  uint64_t uintptr;
+  void     *ptr;
+} ptr_t;
+
+static void *hex_to_ptr(const std::string &hex)
+{
+  ptr_t p;
+
+  assert(hex.length() == sizeof(p) * 2);
+  sscanf(hex.c_str(), "%" SCNx64, &p.uintptr);
+  return p.ptr;
+}
+
+static std::string ptr_to_hex(void *ptr)
+{
+  ptr_t p = { .ptr = ptr };
+  char buf[sizeof(p) * 2 + 1];
+
+  snprintf(buf, sizeof(buf), "%016" PRIx64, p.uintptr);
+  return std::string(buf);
+}
+
+/*
+ * Used for refcounters print on the last context put, almost duplicates
+ * global context refcounter, sigh.
+ */
+static std::atomic<int> ctx_ref(1);
+
+static void create_or_get_ceph_context(struct ceph_msgr_options *o)
+{
+  if (g_ceph_context) {
+    g_ceph_context->get();
+    ctx_ref++;
+    return;
+  }
+
+  boost::intrusive_ptr<CephContext> cct;
+  vector<const char*> args;
+
+  if (o->conffile)
+    args = { "--conf", o->conffile };
+
+  cct = global_init(NULL, args, CEPH_ENTITY_TYPE_CLIENT,
+		    CODE_ENVIRONMENT_UTILITY,
+		    CINIT_FLAG_NO_DEFAULT_CONFIG_FILE);
+  /* Will use g_ceph_context instead */
+  cct.detach();
+
+  common_init_finish(g_ceph_context);
+  g_ceph_context->_conf.apply_changes(NULL);
+}
+
+static void put_ceph_context(void)
+{
+  if (--ctx_ref == 0) {
+    Formatter* f;
+
+    std::cout << ">>>>>>>>>>>>> PERFCOUNTERS BEGIN <<<<<<<<<<<<" <<
```

Comment:
may be use dout instead and don't pollute stdout?

Reply:
That intention was to pollute stdout of fio, i.e. I want to see these counters and then fio report.

Reply:
IMO it's better to have the same reporting model for all FIO plugins hence I'd suggest to use the approach similar to fio-ceph-objectstore plugin. BTW using Ceph logging permits you to dump to stdout (after some tuning).

Reply:
>IMO it's better to have the same reporting model for all FIO plugins
true, will fix then.